test/e2e/tests/test_references.py

```diff
@@ -0,0 +1,163 @@
+
+    assert cr is not None
+    assert k8s.get_resource_exists(ref)
+
+    # NOTE(jaypipes): We specifically do NOT wait for the Role to exist in
+    # the IAM API here because we will create the referred-to Policy and
+    # wait for the reference to be resolved
+
+    yield (ref, cr, role_name)
+
+    if k8s.get_resource_exists(ref):
+        # If all goes properly, we should not hit this because the test cleans
+        # up the child resource before exiting...
+        _, deleted = k8s.delete_custom_resource(
+            ref,
+            period_length=DELETE_ROLE_TIMEOUT_SECONDS,
+        )
+        assert deleted
+
+        role.wait_until_deleted(role_name)
+
+
+@pytest.fixture(scope="module")
+def referred_policy(referred_policy_name):
+    policy_desc = "a referred-to policy"
+
+    replacements = REPLACEMENT_VALUES.copy()
+    replacements['POLICY_NAME'] = referred_policy_name
+    replacements['POLICY_DESCRIPTION'] = policy_desc
+
+    resource_data = load_resource(
+        "policy_simple",
+        additional_replacements=replacements,
+    )
+
+    ref = k8s.CustomResourceReference(
+        CRD_GROUP, CRD_VERSION, POLICY_RESOURCE_PLURAL,
+        referred_policy_name, namespace="default",
+    )
+    k8s.create_custom_resource(ref, resource_data)
+    cr = k8s.wait_resource_consumed_by_controller(ref)
+
+    cr = k8s.get_resource(ref)
+    assert cr is not None
+    assert 'status' in cr
+    assert 'ackResourceMetadata' in cr['status']
+    assert 'arn' in cr['status']['ackResourceMetadata']
+    policy_arn = cr['status']['ackResourceMetadata']['arn']
+
+    policy.wait_until_exists(policy_arn)
+
+    assert cr is not None
+    assert k8s.get_resource_exists(ref)
+
+    yield (ref, cr, policy_arn)
+
+    _, deleted = k8s.delete_custom_resource(
+        ref,
+        period_length=DELETE_POLICY_TIMEOUT_SECONDS,
+    )
+    assert deleted
+
+    policy.wait_until_deleted(policy_arn)
+
+
+@service_marker
+@pytest.mark.canary
+class TestReferences:
+    def test_references(self, referring_role, referred_policy):
+
+        # create the resources in order that initially the reference resolution
+        # fails and then when the referenced resource gets created, then all
+        # resolutions eventually pass and resources get synced.
+        role_ref, role_cr, role_name = referring_role
+
+        time.sleep(1)
+
+        policy_ref, policy_cr, policy_arn = referred_policy
+
+        time.sleep(CHECK_WAIT_AFTER_REF_RESOLVE_SECONDS)
+
+        condition.assert_synced(policy_ref)
+        condition.assert_synced(role_ref)
+
+        role.wait_until_exists(role_name)
+
+        # NOTE(jaypipes): We need to manually delete the Role first because
+        # pytest fixtures will try to clean up the Policy fixture *first*
+        # (because it was initialized after Role) but if we try to delete the
+        # Role before the Policy, the cascading delete protection of resource
+        # references will mean the Role won't be deleted.
```

Comment:
Won't the reconciler retry the deletion until both are deleted? This seems like a nice edge case for references that we can test

Reply:
See the explanation... it's because the fixtures are cleaned up in the reverse order of initialization. And so if we just relied on the test case's teardown to delete (both) the role and policy, it would fail the `assert` in the teardown of the Policy fixture because it would not delete the resource until the Role resource was successfully deleted.

Reply:
Ahhh failing the assertion. Hmm yeah that sucks. Ok

Reply:
How do you feel about combining both YAML documents into a single file and applying them at the same time? Since we aren't asserting anything between creating the two resources lets just do it simultaneously?